File: android/app/src/main/java/dk/betterw4/android/feature/birthdays/BirthdayRepository.kt
package dk.betterw4.android.feature.birthdays

import dk.betterw4.android.core.cache.CachePolicy
import dk.betterw4.android.core.cache.SimpleCache
import dk.betterw4.android.core.cache.W4Surface
import dk.betterw4.android.core.result.AppError
import dk.betterw4.android.core.result.AppResult
import dk.betterw4.android.core.w4.W4Client
import dk.betterw4.android.core.w4.W4Dates
import dk.betterw4.android.core.w4.W4Urls
import dk.betterw4.android.core.w4.model.FetchPriority
import dk.betterw4.android.core.w4.session.SessionController
import dk.betterw4.android.feature.demo.DemoData
import dk.betterw4.android.feature.directory.DirectoryEntityKind
import java.time.LocalDate
import java.time.YearMonth
import java.time.format.DateTimeFormatter
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BirthdayRepository @Inject constructor(
    private val client: W4Client,
    private val cache: SimpleCache,
    private val session: SessionController,
) {
    suspend fun loadMonth(
        ref: BirthdayMonthRef,
        force: Boolean = false,
        priority: FetchPriority = FetchPriority.Important,
    ): AppResult<BirthdayMonth> {
        val student = session.currentStudent ?: return AppResult.Failure(AppError.Unauthorized)
        if (student.isDemo) return AppResult.Success(demoMonth(ref))

        val key = cacheKey(student.studentId, ref)
        if (!force) {
            cache.getWithMeta(key)?.let { cached ->
                if (CachePolicy.isFresh(cached.updatedAtMs, W4Surface.PEOPLE)) {
                    return AppResult.Success(parsed(cached.value, ref))
                }
            }
        }
        return when (
            val res = client.get(
                W4Urls.Routes.BIRTHDAYS_INDEX,
                query = mapOf(
                    "month" to ref.month.toString(),
                    "year" to ref.year.toString(),
                ),
                priority = priority,
            )
        ) {
            is AppResult.Success -> {
                cache.put(key, res.data.body)
                AppResult.Success(parsed(res.data.body, ref))
            }
            is AppResult.Failure -> {
                cache.get(key)?.let { return AppResult.Success(parsed(it, ref)) }
                res
            }
        }
    }

    private fun parsed(html: String, ref: BirthdayMonthRef): BirthdayMonth {
        val parsed = W4BirthdayParser.parse(html)
        return parsed.copy(
            year = parsed.year ?: ref.year,
            month = parsed.month ?: ref.month,
            monthLabel = parsed.monthLabel ?: ref.label,
            previous = parsed.previous ?: ref.offset(-1),
            next = parsed.next ?: ref.offset(1),
        )
    }

    companion object {
        private val DISPLAY_DAY = DateTimeFormatter.ofPattern("EEE d MMM", Locale.UK)

        fun cacheKey(studentId: String, ref: BirthdayMonthRef): String =
            "birthdays_${studentId}_${ref.year}_${ref.month.toString().padStart(2, '0')}"

        fun demoMonth(
            ref: BirthdayMonthRef,
            today: LocalDate = W4Dates.today(),
        ): BirthdayMonth {
            val students = DemoData.directory.filter { it.kind == DirectoryEntityKind.STUDENT }
            val staff = DemoData.directory.filter { it.kind == DirectoryEntityKind.TEACHER }
            val length = YearMonth.of(ref.year, ref.month).lengthOfMonth()
            val isCurrent = today.year == ref.year && today.monthValue == ref.month

            fun person(entity: dk.betterw4.android.feature.directory.DirectoryEntity) = BirthdayPerson(
                uwcId = entity.id,
                name = entity.name,
                isStaff = entity.kind == DirectoryEntityKind.TEACHER,
                photoUrl = entity.avatarUrl,
            )

            fun day(number: Int, people: List<dk.betterw4.android.feature.directory.DirectoryEntity>): BirthdayDay {
                val date = runCatching { LocalDate.of(ref.year, ref.month, number) }.getOrNull()
                return BirthdayDay(
                    date = date,
                    dayNumber = number,
                    dateLabel = date?.format(DISPLAY_DAY) ?: number.toString(),
                    people = people.map(::person),
                )
            }

            val days = buildList {
                if (isCurrent) {
                    add(day(today.dayOfMonth, students.take(1)))
                    val tomorrow = today.dayOfMonth + 1
                    if (tomorrow <= length) {
                        add(day(tomorrow, students.drop(1).take(1)))
                    }
                    val later = minOf(today.dayOfMonth + 6, length)
                    if (later != today.dayOfMonth && later != tomorrow) {
                        add(day(later, staff.take(1) + students.drop(2).take(2)))
                    }
                } else {
                    add(day(3, students.take(2)))
                    add(day(minOf(18, length), staff.take(1)))
                }
            }.sortedBy { it.date ?: LocalDate.MIN }

            return BirthdayMonth(
                monthLabel = ref.label,
                year = ref.year,
                month = ref.month,
                previous = ref.offset(-1),
                next = ref.offset(1),
                days = days,
            )
        }
    }
}

File: ios/BetterW4/BirthdayRepository.swift
//
//  BirthdayRepository.swift
//  BetterW4
//
//  `people/birthdays` for the current month, `people/birthdays/index` with
//  `month` + `year` siblings for any other. Cache: `W4Surface.people`, TTL
//  7 days — who has a birthday on which day does not change during term.
//  "Today" is computed from the Oslo clock, never from a cached CSS class.
//

import Foundation

actor BirthdayRepository {

    static let shared = BirthdayRepository()

    private let client: any W4SecondaryFetching
    private let cache: W4PageCache
    private let resolveContext: @Sendable () throws -> W4RequestContext

    init(
        client: any W4SecondaryFetching = W4HTTPClient(),
        cache: W4PageCache = .shared,
        resolveContext: @escaping @Sendable () throws -> W4RequestContext = {
            try W4RequestContext.require()
        }
    ) {
        self.client = client
        self.cache = cache
        self.resolveContext = resolveContext
    }

    static func cacheKey(_ ref: BirthdayMonthRef) -> String {
        String(format: "birthdays-%04d-%02d", ref.year, ref.month)
    }

    func loadMonth(
        _ ref: BirthdayMonthRef,
        forceRefresh: Bool = false,
        priority: FetchPriority = .important
    ) async throws -> W4Loaded<BirthdayMonth> {
        let context = try resolveContext()
        if context.isDemo {
            return W4Loaded(Self.demoMonth(ref), freshness: .demo)
        }
        let loaded = try await W4SecondaryPageLoader.loadHTML(
            surface: .people,
            key: Self.cacheKey(ref),
            route: W4Routes.R.birthdaysIndex,
            query: [
                "month": String(ref.month),
                "year": String(ref.year)
            ],
            forceRefresh: forceRefresh,
            priority: priority,
            context: context,
            client: client,
            cache: cache
        )
        return loaded.map { html in
            var parsed = W4BirthdayParser.parse(html)
            if parsed.year == nil { parsed.year = ref.year }
            if parsed.month == nil { parsed.month = ref.month }
            if parsed.monthLabel == nil { parsed.monthLabel = ref.label }
            if parsed.previous == nil { parsed.previous = ref.offset(by: -1) }
            if parsed.next == nil { parsed.next = ref.offset(by: 1) }
            return parsed
        }
    }

    func cachedMonth(_ ref: BirthdayMonthRef) async -> W4Loaded<BirthdayMonth>? {
        guard let context = try? resolveContext() else { return nil }
        if context.isDemo {
            return W4Loaded(Self.demoMonth(ref), freshness: .demo)
        }
        let cached = await W4SecondaryPageLoader.cachedHTML(
            surface: .people,
            key: Self.cacheKey(ref),
            context: context,
            cache: cache
        )
        return cached?.map { html in
            var parsed = W4BirthdayParser.parse(html)
            if parsed.year == nil { parsed.year = ref.year }
            if parsed.month == nil { parsed.month = ref.month }
            if parsed.monthLabel == nil { parsed.monthLabel = ref.label }
            if parsed.previous == nil { parsed.previous = ref.offset(by: -1) }
            if parsed.next == nil { parsed.next = ref.offset(by: 1) }
            return parsed
        }
    }

    // MARK: - Demo

    static func demoMonth(_ ref: BirthdayMonthRef, now: Date = TimeProvider.now) -> BirthdayMonth {
        let people = DirectoryRepository.demoPeople
        let students = people.filter { $0.kind == .student }
        let staff = people.filter { $0.kind == .staff }
        let today = W4Dates.startOfDay(now)
        let todayParts = W4Dates.calendar.dateComponents([.year, .month, .day], from: today)
        let length = W4Dates.calendar.range(of: .day, in: .month, for: W4Dates.date(year: ref.year, month: ref.month, day: 1) ?? today)?.count ?? 30

        func person(_ source: DirectoryPerson) -> BirthdayPerson {
            BirthdayPerson(
                uwcId: source.uwcId,
                name: source.displayName,
                isStaff: source.kind == .staff,
                profileRoute: source.profileRoute,
                profileURL: source.profileURL,
                photoURL: source.photoURL
            )
        }

        func day(_ number: Int, people: [DirectoryPerson]) -> BirthdayDay {
            let date = W4Dates.date(year: ref.year, month: ref.month, day: number)
            let label = date.map { displayDayFormatter.string(from: $0) } ?? "\(number)"
            return BirthdayDay(
                date: date,
                dayNumber: number,
                dateLabel: label,
                people: people.map(person)
            )
        }

        var days: [BirthdayDay] = []
        let isCurrent = todayParts.year == ref.year && todayParts.month == ref.month
        if isCurrent, let todayNumber = todayParts.day {
            days.append(day(todayNumber, people: Array(students.prefix(1))))
            let tomorrowNumber = todayNumber + 1
            if tomorrowNumber <= length {
                days.append(day(tomorrowNumber, people: Array(students.dropFirst().prefix(1))))
            }
            let later = min(todayNumber + 6, length)
            if later != todayNumber, later != tomorrowNumber {
                days.append(day(later, people: Array(staff.prefix(1)) + Array(students.dropFirst(2).prefix(2))))
            }
        } else {
            days.append(day(3, people: Array(students.prefix(2))))
            days.append(day(min(18, length), people: Array(staff.prefix(1))))
        }
        days.sort { lhs, rhs in
            if let a = lhs.date, let b = rhs.date { return a < b }
            return lhs.dayNumber < rhs.dayNumber
        }
        return BirthdayMonth(
            monthLabel: ref.label,
            year: ref.year,
            month: ref.month,
            previous: ref.offset(by: -1),
            next: ref.offset(by: 1),
            days: days
        )
    }

    private static let displayDayFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_GB")
        formatter.timeZone = W4Dates.zone
        formatter.calendar = W4Dates.calendar
        formatter.dateFormat = "EEE d MMM"
        return formatter
    }()
}
